### PR TITLE
Enable unmap hack for java 9

### PR DESCRIPTION
**Pull request by rmuir**
This mechanism changed for java 9: requires a "fake" permission of `accessClassInPackage.jdk.internal.ref`

Currently unmapping will not work with java 9 unless we make changes:

```
  2> REPRODUCE WITH: gradle :core:test -Dtests.seed=D11864EB2E98C3EF -Dtests.class=org.elasticsearch.common.lucene.LuceneTests -Dtests.method="testMMapHackSupported" -Des.logger.level=WARN -Dtests.security.manager=true -Dtests.locale=jmc-TZ -Dtests.timezone=Antarctica/DumontDUrville
FAILURE 0.13s J0 | LuceneTests.testMMapHackSupported <<< FAILURES!
   > Throwable #1: java.lang.AssertionError: MMapDirectory does not support unmapping: Unmapping is not supported, because not all required permissions are given to the Lucene JAR file: java.security.AccessControlException: access denied ("java.lang.RuntimePermission" "accessClassInPackage.jdk.internal.ref") [Please grant at least the following permissions: RuntimePermission("accessClassInPackage.sun.misc"), RuntimePermission("accessClassInPackage.jdk.internal.ref"), and ReflectPermission("suppressAccessChecks")]
```

See https://issues.apache.org/jira/browse/LUCENE-6989 for more information. 

I also had to upgrade my gradle to the latest to work at all with java 9, and fix a test so the source code compiles at all. I only ran this small unit test because hotspot on java 9 has been broken for a long time.

### CODE REVIEW

**Comment by uschindler** [comment]
Looks good.

**Comment by mikemccand** [comment]
+1, thanks @rmuir.  It's important we can run under Java 9 ea even if recent builds have been badly broken...

File: core/src/test/java/org/elasticsearch/cluster/routing/PrimaryAllocationIT.java [hunk 1]
```diff
@@ -109,7 +109,12 @@ private void createStaleReplicaScenario() throws Exception {
         logger.info("--> check that old primary shard does not get promoted to primary again");
         // kick reroute and wait for all shard states to be fetched
         client(master).admin().cluster().prepareReroute().get();
-        assertBusy(() -> assertThat(internalCluster().getInstance(GatewayAllocator.class, master).getNumberOfInFlightFetch(), equalTo(0)));
```
Comment (uschindler) [inline core/src/test/java/org/elasticsearch/cluster/routing/PrimaryAllocationIT.java]: was this change wanted, it looks unrelated to this issue?!

Comment (rmuir) [inline core/src/test/java/org/elasticsearch/cluster/routing/PrimaryAllocationIT.java]: as mentioned in the issue, it will not compile with java 9.

Comment (uschindler) [inline core/src/test/java/org/elasticsearch/cluster/routing/PrimaryAllocationIT.java]: OK. That seems to be a bug in javac :-) Maybe we should open issue!

Comment (uschindler) [inline core/src/test/java/org/elasticsearch/cluster/routing/PrimaryAllocationIT.java]: I would have tried:

```java
assertBusy(() -> {
  assertThat(internalCluster().getInstance(GatewayAllocator.class, master).getNumberOfInFlightFetch(), equalTo(0));
});
```

Comment (rmuir) [inline core/src/test/java/org/elasticsearch/cluster/routing/PrimaryAllocationIT.java]: I gave it the obvious fix without any special sugar. and plenty of assertBusy's, actually most of them, are written this way.

again, as it was a compiler failure, the style is unimportant to me really. More important was to get mmap working :)
